frontend/app/helpers/spree/addresses_helper.rb

```diff
@@ -65,9 +65,18 @@ def address_state(form, country, address_id = 'b')
     end
 
     def user_available_addresses
-      return unless try_spree_current_user
+      return [] unless try_spree_current_user
 
-      try_spree_current_user.addresses.where(country: available_countries)
+      checkout_countries = current_store.countries_available_for_checkout
+
+      states = checkout_countries.each_with_object([]) do |country, memo|
+        memo << current_store.states_available_for_checkout(country)
+      end
+
+      try_spree_current_user.addresses.
+        joins(country: :states).
+        where(state: states.flatten).
```

Comment:
maybe instead making here `flatten` during query we could use that in `states` variable and pass flatted array to query ?

Reply:
@przemosk changed! https://github.com/spree/spree/pull/10594/commits/d4f1c0eb821dc70a32710443ea03a1e0637bc97f